### PR TITLE
LPS-119037 Fix clicking on user in mentions list throwing JavaScript error

### PR DESCRIPTION
**Pull request by holatuwol**
[Issue](https://issues.liferay.com/browse/LPS-119037)

Clicking on a user in the mentions list throws javascript error: `prevTriggerPosition.container.split is not a function`

**Step to reproduce:**
- Add a public user (user1)
- Add a Message board portlet to any page
- Add a new thread
- In the body, add a mention ( @user1 )
- Add a whitespace, then add a line break
- Add a whitespace, then press Backspace button to remove that whitespace
- Note the list of users is displayed
- Click on Mention list to add mention of user1 again

### CODE REVIEW

**Comment by jbalsas** [comment]
Thanks for sending this over, @holatuwol! 

Please, for future PRs we would appreciate it if you could take a look at our
- [Frontend Infrastructure Team code review process](https://github.com/liferay/liferay-frontend-guidelines/blob/master/dxp/code_review.md)
- [Pull Request Etiquette](https://github.com/liferay/liferay-frontend-guidelines/blob/master/general/pull_requests.md)

The most relevant parts for our review process would be:
- Always provide a **title**
- Always provide a **description**
- Always provide a **test plan**
- [Optional but extremely useful] Add technical context about why this is the proper solution, what did you try, what did you discard and why...

I know it might seem redundant since some of this information might already be in the linked LPS, but these simple steps really help us reviewing and keeping track of the contributed work throughout the whole workflow. We really appreciate your assistance in this!!

PS: I'll edit the title and description for this one based on the ticket information

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/autocomplete/autocomplete.js [hunk 1]
```diff
@@ -25,6 +25,8 @@
 
 	var STR_SPACE = ' ';
 
+	var BR_TAG = 'BR';
```
Comment (wincent) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/autocomplete/autocomplete.js:28]: Should probably sort this above `CSS_LFR_AC_CONTENT`. The list of declarations here isn't entirely sorted, but it mostly is, so we should try to keep it that way.

**Comment by wincent** [comment]
Other than the sorting nit this change looks fine to me, although I am taking it on trust (I haven't taken the time to fully understand the hundreds of lines in this file).

Going to fire of tests for verification.

**Comment by wincent** [comment]
ci:test:sf

**Comment by wincent** [comment]
ci:test:relevant

**Comment by wincent** [comment]
ci:test:bundle

**Comment by wincent** [comment]
Test failure looks spurious to me, so going to retry:

> ```
> no such session
>  (Driver info: chromedriver=2.37.544315 (730aa6a5fdba159ac9f4c1e8cbc59bf1b5ce12b7),platform=Linux 4.14.138-rancher x86_64) (WARNING: The server did not provide any stacktrace information)
> ```

**Comment by wincent** [comment]
ci:test:relevant

**Comment by wincent** [comment]
So I tested this manually and I am not sure what the expected behavior is.

1. Add a public user (user1)
2. Add a Message board portlet to any page
3. Add a new thread
4. In the body, add a mention ( @user1 )
5. Add a whitespace, then add a line break

Screenshot 2020-09-04 -154056-BC7XVMfG@2x

At this point, I see the user list... Why?

Anyway, I continue with the repro recipe:

6. Add a whitespace (at this point the menu hides), then press Backspace button to remove that whitespace
7. _Note the list of users is displayed_ (nope, I see no menu at this point).
8. Click on Mention list to add mention of user1 again (can't do this because there is no menu)

So, if I Backspace to previous line, then hit Enter to open a new line again, menu shows and when I click on it I see this JS error:

Screenshot 2020-09-04 -154135-2Ix1zseV@2x

All of this with the bundle I downloaded from `ci:test:bundle`.

Seems like this might need something more @holatuwol?

**Comment by holatuwol** [comment]
I see, thanks for the due diligence.

**Comment by binhtran92** [comment]
Hi @holatuwol and @wincent 
I tested it again and It does works. 
I feel like the `autocomplete.js` is cached. 
Please help me to make sure on your browser the `autocomplete.js` is updated.
Thank you.

**Comment by dinh-thai** [comment]
Hi @holatuwol and @wincent ,
I tested it on the bundle from ci:test:bundle and it does works.
I attack my video evidence at ticket [LPP-38588](https://issues.liferay.com/browse/LPP-38588)
Same as Binh comment, Please help me to make sure all cached data are cleared before test this issue..
Thank you.
Dinh Thai

**Comment by wincent** [comment]
> Please help me to make sure all cached data are cleared before test this issue..

I'll re-download the bundle test this again. However, I am at a loss to explain why a cached artifact would explain the behavior seen in my screenshots; as you can see, my browser won't cache anything with the devtools open (which they are in the screenshot), and it seems odd to me that a brand new `bundles` directory could have stale state in it. 🤷 

Screenshot 2020-09-07 -102454-pnEAatZ4@2x

**Comment by wincent** [comment]
Ok, on retesting this, I no longer see the JS error. However, I still see this:

> 5\. Add a whitespace, then add a line break
> 
> Screenshot 2020-09-04 -154056-BC7XVMfG@2x
> 
> At this point, I see the user list... Why?

Can anybody comment on why the mentions list pops up at this point? It feels to me like a bug.

Video:

Screenshot 2020-09-07 -105235-HpgYkFp6

That's the result of:

- Typing `Hello @user`
- Clicking the mentions menu (would be nice to tab to inserted the selected menu like mention inputs work on most popular platforms, but seems that that's not a feature here).
- Pressing space, then enter.

It doesn't seem to make much sense that the mentions menu would appear again at this point, but it does.

**Comment by jbalsas** [comment]
> It doesn't seem to make much sense that the mentions menu would appear again at this point, but it does.

Yeah, I agree. I think this is partly the goal of the PR, to fix the autocompletion menu from appearing in the wrong place. It feels like the fix addresses the `backspace` behaviour but not this one, so it'd be good if we can get it all fixed at once.

@dinh-thai, @binhtran92, can you verify the steps @wincent is providing? This can be either caused by this PR or already present, but in both cases we should probably try to account for it rather than patch it independently. Otherwise, we might be adding isolated patch after patch making this harder and harder to maintain.

On a related note, @carloslancha attempted migrating to the ootb [autocomplete](https://ckeditor.com/docs/ckeditor4/latest/features/autocomplete.html) plugin. This proved a tad bit more complicated than expected, so we had to postpone it to after the `7.3` release but I think it should do for a more robust behaviour coming forward. 

This last thing was to say that I guess it's okay to patch this and try to replace it with the default plugin as soon as we can.

**Comment by dinh-thai** [comment]
Hi @wincent 
I have upload video evidence on LPP-38588 because I cannot upload video here. Can you goto [LPP-38588](https://issues.liferay.com/browse/LPP-38588) to watch my video? I ensure that this issue is fixed.
As I see on your screen, it's not be load the new autocomplete.js file. 
When you go to the page, can you clear cache as I attach image.
empty_cache

Thanks
Dinh Thai

**Comment by wincent** [comment]
> Can you goto [LPP-38588](https://issues.liferay.com/browse/LPP-38588) to watch my video?

Screenshot 2020-09-07 -111119-J7SnPPBc@2x

Seems I can't. 75KB does seem a bit on the small side.

Thanks for the tip about emptying the cache and hard-reloading though.

Did you read my comment? I'll emphasize the important bit:

> Ok, on retesting this, **I no longer see the JS error.**

The bit about the mention popping up unexpectedly still definitely occurs, even when I do the "empty cache" trick:

ScreenFlow

**Comment by dinh-thai** [comment]
Hi @wincent ,
I discovered this issue during I fix LPP-38497. And your screen now is same the issue present on Master now.
After fixed this issue, I test it is fixed both git local branch and bundle from ci:test:bundle. I don't know why it doesn't work with you. I think I will discuss with @holatuwol about this.
And I ensure that in this branch, this issue is fixed and you're issue is no longer present. (as in my video evidence, but you cannot watch)
Thank you,
Dinh Thai

**Comment by dinh-thai** [comment]
> Ok, on retesting this, I no longer see the JS error.

About this, this issue is not present when the first time you click on that row.
When you press Enter button, the mention list is displayed again at the empty row, and now if you click on mention list, js error will display.
This is the issue of this ticket: **mention list display at empty row and js error display when click on mention list.**
My solution is do not display mention list at empty row.
Thank you,
Dinh Thai

**Comment by wincent** [comment]
Yeah, I am at a loss to describe what is going on. I tested this in an incognito window to rule out browser issues, and tested on every browser I have on my machine (not counting VMs), which basically means a bunch of Chrome/Chromium variants (Chrome, Chrome Canary, Chromium, Opera, Edge) + Firefox. The only browser in which I don't see the behavior is Firefox. Actually, forgot about Safari: I don't see it there, but the behavior is subtly different — when I click the mention in the list, the cursor disappears, but the text area still has focus and you can continue typing.

There's obviously something anomalous on my machine, or it could be that the bundle I have is yielding different results than I'd get if I applied this PR on top of `master`. @jbalsas does not see the behavior I am seeing with the PR on top of current `master`, but he says he does see it if he dismisses the mentions list with `Escape` and then hits `Enter` (I can repro that too).

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/autocomplete/autocomplete.js [hunk 2]
```diff
@@ -271,6 +273,16 @@
...
+
+			if (
+				query &&
+				prevTriggerPosition.container.$.lastElementChild &&
+				prevTriggerPosition.container.$.lastElementChild.nodeName ===
+					BR_TAG
+			) {
+				query = null;
+			}
+
```
Comment (markocikos) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/autocomplete/autocomplete.js:285]: We can simplify this block a bit, but it's not a blocker:

```js
var query = null;

if (
	prevTriggerPosition.container.$.lastElementChild?.nodeName !==
	BR_TAG
) {
	query = prevTriggerPosition.query;
}
```

**Review by markocikos — APPROVED** [review]
The solution works for me:

autocomplete

**Comment by wincent** [comment]
> The solution works for me:

@markocikos: Can you repro this in Firefox?

> if he dismisses the mentions list with Escape and then hits Enter (I can repro that too).

**Comment by markocikos** [comment]
> @markocikos: Can you repro this in Firefox?

Yep, I can reproduce the issue on Firefox, by using escape. For some weird reason, now I can only reproduce in in Chrome using escape. I am sure I was able to reproduce it using backspace a minute ago.

Anyways, this PR does seem to fix the issue on both browsers (after it was triggered by escape).

There is also a bug in Firefox, where you need to hit backspace eight times to go up to previous row. That's also on master, so it's not related to this PR.

**Comment by wincent** [comment]
> There is also a bug in Firefox, where you need to hit backspace eight times to go up to previous row. That's also on master, so it's not related to this PR.

While it would be good to fix the JS error soon (because JS errors are bad), it does feel a bit like whack-a-mole where we patch over one issue in some browsers, while another issue or two manifests (or not) in different browsers.

We can either ship this patch (and then more follow-ups to fix issues) or try to make this whole thing more solid in a more definitive way. Given that [LPS-119037](https://issues.liferay.com/browse/LPS-119037) has [LPP-38588](https://issues.liferay.com/browse/LPP-38588) attached, I guess there is _some_ urgency here. I'd be ok with shipping it as-is to address the immediate pain point, but at some point we have to make this thing work better. Which I guess is what @jbalsas said earlier:

> This last thing was to say that I guess it's okay to patch this and try to replace it with the default plugin as soon as we can.

**Comment by holatuwol** [comment]
Just double-checking, is the frontend team working out what would be the best path forward, or are we likely to move forward with the pull request as-is?

**Comment by wincent** [comment]
Sorry for the delay @holatuwol. I wanted to allow some time for people to respond. Given that nobody has, I think we do what we've said in the previous comments:

> This last thing was to say that I guess it's okay to patch this and try to replace it with the default plugin as soon as we can. (@jbalsas)

> We can simplify this a bit, but it's not a blocker (@markocikos)

> I'd be ok with shipping it as-is to address the immediate pain point, but at some point we have to make this thing work better. (@wincent)

So, let's forward as-is and look at doing follow-ups for the remaining issues.

**Comment by wincent** [comment]
ci:forward

**Pull request closed without merging**